clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -110,77 +113,70 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
                 // satisfy the request
                 freeUp(size);
                 this.availableMemory -= size;
-                lock.unlock();
                 return allocateByteBuffer(size);
```

Comment:
If an OOM is thrown from here, it seems the available memory would be already decremented and not added back.